packages/medical-domain/src/domains/repositories/medical/medical-data.service.ts

```diff
@@ -355,11 +266,132 @@ class MedicalDataService {
+        console.log({ message, rawData: raw })
       }
     })
   }
 
+  add(data: MedicalData): void {
+    if (data.bolus) {
+      this.medicalData.bolus = this.medicalData.bolus.concat(data.bolus)
+    }
+    if (data.basal) {
+      this.medicalData.basal = this.medicalData.basal.concat(data.basal)
+    }
+    if (data.cbg) {
+      this.medicalData.cbg = this.medicalData.cbg.concat(data.cbg)
+    }
+    if (data.confidentialModes) {
+      this.medicalData.confidentialModes = this.medicalData.confidentialModes.concat(data.confidentialModes)
+    }
+    if (data.deviceParametersChanges) {
+      this.medicalData.deviceParametersChanges = this.medicalData.deviceParametersChanges.concat(data.deviceParametersChanges)
+    }
+    if (data.messages) {
+      this.medicalData.messages = this.medicalData.messages.concat(data.messages)
+    }
+    if (data.meals) {
+      this.medicalData.meals = this.medicalData.meals.concat(data.meals)
+    }
+    if (data.physicalActivities) {
+      this.medicalData.physicalActivities = this.medicalData.physicalActivities.concat(data.physicalActivities)
+    }
+    if (data.pumpSettings) {
+      this.medicalData.pumpSettings = this.medicalData.pumpSettings.concat(data.pumpSettings)
+    }
+    if (data.reservoirChanges) {
+      this.medicalData.reservoirChanges = this.medicalData.reservoirChanges.concat(data.reservoirChanges)
+    }
+    if (data.smbg) {
+      this.medicalData.smbg = this.medicalData.smbg.concat(data.smbg)
+    }
+    if (data.warmUps) {
+      this.medicalData.warmUps = this.medicalData.warmUps.concat(data.warmUps)
+    }
+    if (data.wizards) {
+      this.medicalData.wizards = this.medicalData.wizards.concat(data.wizards)
+    }
+    if (data.zenModes) {
+      this.medicalData.zenModes = this.medicalData.zenModes.concat(data.zenModes)
+    }
+    if (data.alarmEvents) {
+      this.medicalData.alarmEvents = this.medicalData.alarmEvents.concat(data.alarmEvents)
+    }
+
+    this.deduplicate()
+    this.join()
+    this.setTimeZones()
+    this.group()
+    this.setEndPoints()
+    this.generateFillData()
+    this.generateBasicsData()
+  }
+
+  hasDiabetesData(): boolean {
+    return (
+      this.medicalData.basal.length > 0 ||
+      this.medicalData.bolus.length > 0 ||
+      this.medicalData.cbg.length > 0 ||
+      this.medicalData.smbg.length > 0 ||
+      this.medicalData.wizards.length > 0
+    )
+  }
+
+  filterByDate(epochStart: number, epochEnd: number): Datum[] {
+    return this.data.filter((d) => {
+      const datumEpoch = this.getDatumEpoch(d)
+      return datumEpoch > epochStart && datumEpoch < epochEnd
+    })
+  }
+
+  // Basics data are needed to display the cartridge changes in dashboard and in print actions
+  // We may only need to generate these for the current day displayed in dashboard view and on print...
+  generateBasicsData(startDate: string | null = null, endDate: string | null = null): BasicData | null {
```

Comment:
This whole function is tricky but as you didn't write it, I'm ok with this :slightly_smiling_face: